### ISSUE TITLE
Move from per-driver features to per-db (have supports? take a database param as well)

### ISSUE THREAD

**Issue by sbelak** [metabase/metabase#13136]
Currently our MBQL compiler dispatches on driver where driver is a keyword denoting the db. This is somewhat limiting for two reasons:
1) For some DBs (eg. MySQL 5.x vs Maria/8.x) there is a huge gap between in supported features across DB versions forcing us to the lowest common denominator 

2) it assumes a static set of features. It would be nice if we could also indicate presence of extensions (eg. in case of Druid use better sketch algorithms if available) and system/session settings (eg. start of week. See #13090). 

I propose the `driver` parameter we pass around and dispatch on becomes an object that carries with it all supported features and settings.

**Comment by camsaul** [comment]
I think we can start by tweaking the features list multimethods to dispatch off of driver and database, and for JDBC at least we can use `DatabaseMetaData` methods to get the version number.

**Comment by camsaul** [comment]
I think actually we should just make a new `database-supports?` multimethod and have it call `supports?` by default, and deprecated `supports?`. Then we give people a few versions to migrate over to the new method

```clj
(defmulti ^:deprecated supports?
  {:arglists '([driver feature])}
  (fn [driver feature]
    (when-not (driver-features feature)
      (throw (Exception. (tru "Invalid driver feature: {0}" feature))))
    [(dispatch-on-initialized-driver driver) feature])
  :hierarchy #'hierarchy)

(defmulti database-supports?
  {:arglists '([driver feature database])}
  (fn [driver feature _]
    (when-not (driver-features feature)
      (throw (Exception. (tru "Invalid driver feature: {0}" feature))))
    [(dispatch-on-initialized-driver driver) feature])
  :hierarchy #'hierarchy)

(defmethod database-supports? :default
  [driver feature _]
  (supports? driver feature))
```

And then just also record database version number on sync so it's easy to determine whether the feature is supported

**Comment by camsaul** [comment]
@howonlee implemented this in #13136
